src/crimson/os/seastore/transaction_manager.h

```diff
@@ -189,6 +189,17 @@ class TransactionManager : public ExtentCallbackInterface {
 #endif
 	return extent->template cast<T>();
       });
+    } else if (!pin->can_read()) {
+      return lba_manager->to_readable_mapping(t, std::move(pin)
```

Comment:
See comment above, I think the below form is straightforward:
```c++
if (pin->is_indirect()) {
  assert(!v.has_child());
  return lba_manager->to_direct_mapping(t, std::move(pin)  <only accept indirect pin>
  ...
} else if (v.has_child()) {
...
```

Also add comment to the read_pin interface: If the data pin is indirect, the returned extent might contain a larger range of data, also refer to the comment in btree_lba_manager.h about BtreeLBAMapping.

Reply:
Indirect mappings can have children. And indirect and no-child mappings is not necessarily an "unreadable" mapping, it might be a "readable" mapping with no child in memory.

The term "readable" mapping means a mapping through which users can get the child logical extent if that extent is in memory; while "unreadable" mappings are those through which users can not get the child logical extent no matter whether that extent is in memory.

All mappings returned by `TransactionManager::clone_pin()` are "unreadable", because LBAManager doesn't query the direct mappings during cloning.

If a "readable" mapping doesn't have a child, it means that child is not loaded yet and users should directly getting that extent from disk; while, for "unreadable" mappings, users should always find the direct mappings first.

So we can not determine whether an indirect mapping is "unreadable" just by it not having a child, which is why I added the "readable" field to BtreeLBAMapping.

But the semantic of this interface is somewhat confusing indeed, I'll try to make it clear.

Reply:
> Indirect mappings can have children.

@xxhdx1985126  Do you mean that an indirect mapping can be linked to the cloned & cached & stable logical cached extent directly?

If yes, I think we currently should not do that. It looks to break `TM::rewrite_logical_extent()` because the indirect pin cannot be updated.

Even if we can find a way to update the indirect pin, how to proper make rewrite and clone transactions conflict each other will still be an issue, right ?

Reply:
> @xxhdx1985126 Do you mean that an indirect mapping can be linked to the cloned & cached & stable logical cached extent directly?

Yes, this can happen. All indirect mappings returned by lba manager's get_mapping(s) interface are "followable"

> 
> If yes, I think we currently should not do that. It looks to break `TM::rewrite_logical_extent()` because the indirect pin cannot be updated.

This doesn't matter, because "TM::rewrite_logical_extent()" won't meet indirect mappings, because logical extents' laddr are always the keys of direct mappings. Keys of indirect mappings only exists in indirect mappings themselves, and won't affect other components of SeaStore.

Reply:
Yeah, looks these transactions will conflict each other, and only one will be successful.